### PR TITLE
NLVR language fixes

### PR DESCRIPTION
**Pull request by pdasigi**
This PR includes several changes to the NLVR language:
1) I removed lambda functions from the language. This made the box filters less expressive, but the output space is more restricted. The reduced expressivity is not really a problem though. I tried to annotate a random set of 25 sentences, of which I could not annotate only 2 due to the lack of lambdas.
2) I converted each function that had placeholder types into multiple functions where the types of the arguments are unique. This increased the number of terminals, but the grammar now disallows unintended argument types that placeholders earlier allowed.
3) Related to the above change, I removed the attribute functions: "count", "color" and "shape", and pulled them into the assertion and filtering functions. This again increased the number of terminals, but the grammar now requires that the first argument of assertions and filters be sets of entities , and the last arguments be constants. Due to this change, the grammar disallows  trivial contradictions or tautologies like `(assert_greater 3 2)`, and ensures that one of the arguments is really a constant (earlier the decoder was also generating logical forms like `(assert_greater (count (object_in_box all_boxes)) (count (color (object_in_box all_boxes))))`).

After these changes, the number of terminal productions went up from 56 to 101, but the number of non-terminals went down from 48 to 26, and the total number of productions remained unchanged at 121. The average branching factor (number of productions per non-terminal) went up from 2.5 to 4.6, and the maximum branching factor is unchanged at 23 (due to object filtering functions like `square`, `above`, `yellow` etc.). When trained on 1000 questions, all the logical forms produced by the decoder, both during training and validation were valid executable sequences.

### CODE REVIEW

File: tests/data/semparse/worlds/nlvr_world_test.py [hunk 1]
```diff
@@ -11,82 +12,149 @@ def setUp(self):
...
+        # When the objects are from different boxes, top and bottom should return objects from
+        # respective boxes.
+        world = self.custom_world
+        # There are triangles in two boxes, so top should return the top objects from both boxes.
+        assert world.execute("(object_count_equals (top (object_in_box (member_shape_any_equals "
+                             "all_boxes shape_triangle))) 2)")
+
+    def test_count_with_all_equals_throws_execution_error(self):
+        world = self.custom_world
+        with self.assertRaises(ExecutionError):
```
Comment (matt-gardner) [inline tests/data/semparse/worlds/nlvr_world_test.py]: Why does this throw an error?

File: tests/data/semparse/worlds/nlvr_world_test.py [hunk 2]
```diff
@@ -11,82 +12,149 @@ def setUp(self):
...
+            world.execute("(object_count_all_equals (top (object_in_box (member_shape_any_equals "
+                          "all_boxes shape_triangle))) 2)")
+
+    def test_shape_with_equals_throws_execution_error(self):
+        world = self.custom_world
+        with self.assertRaises(ExecutionError):
```
Comment (matt-gardner) [inline tests/data/semparse/worlds/nlvr_world_test.py]: This one too.

File: allennlp/data/semparse/worlds/nlvr_world.py [hunk 3]
```diff
@@ -32,19 +34,26 @@ class Object:
...
 
     def __str__(self):
-        return "%s %s at (%d, %d)" % (self.color, self.shape, self.x_loc, self.y_loc)
+        if self.size == 10:
+            size = "small"
+        elif self.size == 20:
+            size = "medium"
+        else:
+            size = "big"
+        return "%s %s %s at (%d, %d)" % (size, self.color, self.shape, self.x_loc, self.y_loc)
```
Comment (matt-gardner) [inline allennlp/data/semparse/worlds/nlvr_world.py]: You could consider using the `f"{size} {self.color} ..."` syntax from python 3.6.

File: allennlp/data/semparse/worlds/nlvr_world.py [hunk 4]
```diff
@@ -92,17 +101,39 @@ class NlvrWorld(World):
...
+                                  "lesser_equals": operator.le}
+
+        self._set_unary_operators = {"same": self._same,
+                                     "different": self._different}
+
+        self._set_binary_operators = {"all_equals": self._all_equals,
+                                      "any_equals": self._any_equals,
+                                      "none_equals": self._none_equals}
+
+        self._count_functions = {"count": self.count,  # type: ignore
```
Comment (matt-gardner) [inline allennlp/data/semparse/worlds/nlvr_world.py]: I am confused about why some of these functions have underscores and some of them don't.

File: allennlp/data/semparse/worlds/nlvr_world.py [hunk 5]
```diff
@@ -189,130 +239,218 @@ def execute(self, logical_form: str) -> bool:
...
-        """
-        Acceptable syntax is count(object_set).
-        """
-        arguments_list = sub_expression[1]
-        if arguments_list[0].startswith("filter_") or arguments_list[0] == "all_boxes":
-            return self.count(self._execute_box_filter(arguments_list))
+            target_attribute = self._execute_constant(sub_expression[2])
+            # If the length of ``function_name_parts`` is 3, getting the attribute and comparison
+            # operator is easy. However, if it is greater than 3, we need to determine where the
+            # attribute function stops and where the comparison operaot begins.
```
Comment (matt-gardner) [inline allennlp/data/semparse/worlds/nlvr_world.py]: `s/operaot/operator/`

File: allennlp/data/semparse/worlds/nlvr_world.py [hunk 6]
```diff
@@ -189,130 +239,218 @@ def execute(self, logical_form: str) -> bool:
...
+
+        Note that the first kind is a special case of the second where the attribute type is
+        ``count``, comparison operator is ``greater_equals`` and the constant is ``1``.
         """
         # TODO(pradeep): We may want to change the order of arguments here to make decoding easier.
         assert isinstance(sub_expression, list), "Invalid assertion expression: %s" % sub_expression
         if len(sub_expression) == 1 and isinstance(sub_expression[0], list):
             return self._execute_assertion(sub_expression[0])
-        assert isinstance(sub_expression[0], str) and sub_expression[0].startswith("assert_"),\
+        is_assert_function = sub_expression[0].startswith("object_") or \
```
Comment (matt-gardner) [inline allennlp/data/semparse/worlds/nlvr_world.py]: This is a lot of complex logic.  Would it be simpler to just define a single function for each of your terminal functions, and call it?  Seems likely.  This also has the advantage of being much closer to the eventual use case of "you just write a python API, and we'll parse to that".

File: allennlp/data/semparse/worlds/nlvr_world.py [hunk 7]
```diff
@@ -189,130 +239,218 @@ def execute(self, logical_form: str) -> bool:
...
+            return number_comparison(returned_count, target_attribute)
 
     def _execute_box_filter(self, sub_expression: Union[str, List]) -> Set[Box]:
         """
         Box filtering functions either apply a filter on a set of boxes and return the filtered set,
         or return all the boxes.
         The elements should evaluate to one of the following:
-            box_filtering_function(set_to_filter, attribute_function, constant)
-            all_boxes
+        ``(box_filtering_function set_to_filter constant)`` or
```
Comment (matt-gardner) [inline allennlp/data/semparse/worlds/nlvr_world.py]: Same comment here - what's stopping you from just having a bunch of methods called `member_count_equals`, etc.?

File: allennlp/data/semparse/worlds/nlvr_world.py [hunk 8]
```diff
@@ -189,130 +239,218 @@ def execute(self, logical_form: str) -> bool:
...
             all_objects
         """
-        if sub_expression[0] == "negate_filter":
-            original_filter_name = sub_expression[1]
-            if isinstance(original_filter_name, list):
-                original_filter_name = original_filter_name[0]
-            original_filter = getattr(self, original_filter_name)
-            initial_set = self._execute_object_filter(sub_expression[2])
-            return self.negate_filter(original_filter, initial_set)
+        if sub_expression[0][0] == "negate_filter":
```
Comment (matt-gardner) [inline allennlp/data/semparse/worlds/nlvr_world.py]: And here.

File: allennlp/data/semparse/worlds/world.py [hunk 9]
```diff
@@ -26,6 +26,20 @@ def __str__(self):
         return repr(self.message)
 
 
+class ExecutionError(Exception):
+    """
+    This exception gets raised when you're trying to execute a logical form that your executor does
+    not understand. This may be because your logical form contains an function with an invalid name
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/data/semparse/worlds/world.py]: Done.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 10]
```diff
@@ -174,7 +175,7 @@ def _check_denotation(best_action_sequence: List[str],
             denotation = world.execute(logical_form)
             denotation_is_correct = str(denotation).lower() == label.lower()
             return True, denotation_is_correct
-        except (RuntimeError, AssertionError, TypeError):
+        except (ParsingError, ExecutionError):
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Yes, I can. Done.

File: tests/data/semparse/worlds/nlvr_world_test.py [hunk 11]
```diff
@@ -11,82 +12,149 @@ def setUp(self):
         test_filename = "tests/fixtures/data/nlvr/sample_data.jsonl"
         data = [json.loads(line)["structured_rep"] for line in open(test_filename).readlines()]
         self.worlds = [NlvrWorld(rep) for rep in data]
+        custom_rep = [[{"y_loc": 21, "size": 20, "type": "triangle", "x_loc": 27, "color":
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply tests/data/semparse/worlds/nlvr_world_test.py]: Done.

File: tests/data/semparse/worlds/nlvr_world_test.py [hunk 12]
```diff
@@ -11,82 +12,149 @@ def setUp(self):
...
-        logical_form = "(assert_greater_equals \
-                         (count (filter_equals all_boxes (lambda x (count (blue (object_in_box (var x))))) 0)) \
-                         1)"
+        logical_form = "(box_exists (member_color_none_equals all_boxes color_blue))"
         # Should evaluate to False.
         assert not nlvr_world.execute(logical_form)
 
+    def test_logical_form_with_box_filter_within_object_filter_executes_correctly(self):
+        nlvr_world = self.worlds[2]
+        # Utterance is "There are atleast three blue items in boxes with blue items" and label
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply tests/data/semparse/worlds/nlvr_world_test.py]: Done.

File: tests/data/semparse/worlds/nlvr_world_test.py [hunk 13]
```diff
@@ -11,82 +12,149 @@ def setUp(self):
...
+        # is "True".
+        logical_form = "(object_count_greater_equals \
+                            (object_in_box (member_color_any_equals all_boxes color_blue)) 3)"
+        # Should evaluate to True.
+        assert nlvr_world.execute(logical_form)
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply tests/data/semparse/worlds/nlvr_world_test.py]: Agreed. Done.

File: tests/data/semparse/worlds/nlvr_world_test.py (hunk 1 above)
Comment (pdasigi) [reply tests/data/semparse/worlds/nlvr_world_test.py]: "*_all_equals", "*_any_equals", "*_none_equals" are defined for sets and compare the elements in a set against a single value. Added a comment for this.

File: tests/data/semparse/worlds/nlvr_world_test.py (hunk 2 above)
Comment (pdasigi) [reply tests/data/semparse/worlds/nlvr_world_test.py]: Added a comment here too.

File: allennlp/data/semparse/worlds/nlvr_world.py (hunk 3 above)
Comment (pdasigi) [reply allennlp/data/semparse/worlds/nlvr_world.py]: Considered and changed. :)

File: allennlp/data/semparse/worlds/nlvr_world.py (hunk 4 above)
Comment (pdasigi) [reply allennlp/data/semparse/worlds/nlvr_world.py]: They should all have underscores. Fixed it.

File: allennlp/data/semparse/worlds/nlvr_world.py (hunk 5 above)
Comment (pdasigi) [reply allennlp/data/semparse/worlds/nlvr_world.py]: Done.

File: allennlp/data/semparse/worlds/nlvr_world.py (hunk 6 above)
Comment (pdasigi) [reply allennlp/data/semparse/worlds/nlvr_world.py]: Yeah, that should be possible. I'll add a TODO for this and come back to this later.

**Pull request merged**